components/ide/jetbrains/backend-plugin/src/main/kotlin/io/gitpod/jetbrains/remote/latest/GitpodPortForwardingService.kt

```diff
@@ -0,0 +1,136 @@
+// Copyright (c) 2022 Gitpod GmbH. All rights reserved.
+// Licensed under the GNU Affero General Public License (AGPL).
+// See License-AGPL.txt in the project root for license information.
+
+package io.gitpod.jetbrains.remote.latest
+
+import com.intellij.openapi.components.service
+import com.intellij.openapi.diagnostic.thisLogger
+import com.intellij.openapi.project.Project
+import com.intellij.remoteDev.util.onTerminationOrNow
+import com.intellij.util.application
+import com.jetbrains.codeWithMe.model.RdPortType
+import com.jetbrains.rd.platform.util.lifetime
+import com.jetbrains.rd.util.lifetime.LifetimeStatus
+import com.jetbrains.rdserver.portForwarding.ForwardedPortInfo
+import com.jetbrains.rdserver.portForwarding.PortForwardingManager
+import com.jetbrains.rdserver.portForwarding.remoteDev.PortEventsProcessor
+import io.gitpod.jetbrains.remote.GitpodManager
+import io.gitpod.jetbrains.remote.GitpodPortsService
+import io.gitpod.supervisor.api.Status
+import io.gitpod.supervisor.api.StatusServiceGrpc
+import io.grpc.stub.ClientCallStreamObserver
+import io.grpc.stub.ClientResponseObserver
+import io.ktor.utils.io.*
+import java.util.concurrent.CompletableFuture
+import java.util.concurrent.TimeUnit
+
+@Suppress("UnstableApiUsage")
+class GitpodPortForwardingService(private val project: Project) {
+    companion object {
+        const val FORWARDED_PORT_LABEL = "gitpod"
+    }
+
+    private val portsService = service<GitpodPortsService>()
+
+    init { start() }
+
+    private fun start() {
+        if (application.isHeadlessEnvironment) return
+
+        observePortsListWhileProjectIsOpen()
+    }
+
+    private fun observePortsListWhileProjectIsOpen() = application.executeOnPooledThread {
+        while (project.lifetime.status == LifetimeStatus.Alive) {
```

Comment:
We don't need `while` and `sleep` here since use observe from grpc below. Just observe one time, and it'll keep respond data.

See also https://github.com/gitpod-io/gitpod/blob/151ed2c0095136f8c604def16b819334378bb3be/components/supervisor/pkg/supervisor/services.go#L206-L210

Reply:
🤔 That `while` has a different function. It's there to cover the case where the user loses connection temporarily. So while the IDE is opened, it should keep trying to reconnect to Supervisor.

Do you think we can conver this case using only PortStatus?
But even so, we'd need to keep a cooldown before trying to connect again, so I think 1sec interval is ok.
We've been using this logic in all other classes (from JetBrains plugin) that requires waiting for reconnection.